extensions/ql-vscode/src/variant-analysis/sarif-processing.ts

```diff
@@ -168,6 +168,16 @@ export function tryGetRule(
   return undefined;
 }
 
+function getFilePath(
+  physicalLocation: sarif.PhysicalLocation,
+): string | undefined {
+  const filePath = physicalLocation.artifactLocation?.uri;
+  if (filePath === undefined || filePath === "" || filePath === "file:/") {
```

Comment:
It seems like according to the SARIF spec, the `file:` syntax is completely valid. Since we only need to support files that are produced by CodeQL, I think we should:
- If there's no scheme (i.e. no `:` present), then we should just return that
- If there is a `file` scheme, we should strip that off and return that
- If there's a different scheme, we should return `undefined` since we don't know what it should link to

For the first two cases, we can check that the string is not empty after removing all `/`, so that it will also return `undefined` for `//`.

I think this makes us pretty compatible with anything that can be produced by CodeQL and minimizing the number of false positives/negatives, but feel free to choose something different.

Reply:
That's very similar to the plan I was considering too, and I think it's a good compromise between being compatible enough and not going too far down the rabbit hole.

Reply:
Upon looking at this again, I'm not sure how or whether we should handle `file:` URIs at all. The problem I'm hitting is that I just don't think it's correct to strip off the scheme and all leading slashes and then treat it like a relative path. Correct me if I'm wrong but I believe that URIs with relative paths just don't exist, or at least aren't part of the RFC spec. This is problematic because what we do with the file path is append it to a "file link prefix" which then constructs a HTTP URL to the file on github.com.

So if we receive a full URI I don't think we can safely convert it into a github link. In which case should we be handling them at all? As far as I can tell, CodeQL is only outputting this because it doesn't have enough data, so this is essentially an error case that we want to suppress. Judging by our parsing code and the fact it never fails otherwise, it appears CodeQL doesn't generate `file:` URIs for correct files.

Since we only need to handle output from CodeQL and not the full SARIF spec, I think that makes it acceptable to reject anything that starts with any kind of URI scheme. Does that make sense?

Reply:
Good point, we're only using this for linking to the file on GitHub so handling any of those should not be necessary. It is possible to make  a relative URI by using `file://my/path` instead of `file:///my/path` (2 vs 3 leading slashes), but that's probably not used that often. I think it would indeed be valid to not support `file:` at all.

Reply:
> It is possible to make a relative URI by using `file://my/path` instead of `file:///my/path` (2 vs 3 leading slashes), but that's probably not used that often.

As you say, this isn't strictly relevant, since we don't expect CodeQL to generate these, but I've found conflicting information online, so I thought I'd try to research. This is mostly for my own future information but it may be of interest to others.

I checked [URI RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986#section-3.3). There are a number of different RFCs for URI but I think this one is still the most recent one from the IETF 🤷 

It says you can have relative paths, but the syntax would be `scheme:path/to/resource` where the path does not start with a slash. You only get `://` when the URI has an authority part. You get this all the time with HTTP URLs because the hostname (e.g. `github.com`) is not part of the path, and the `//` is what denotes this. You also get this in file URIs with `file://localhost/path` but then host can be omitted specifically when it is `localhost` and this leaves you with the triple slash. According to the spec a double slash preceding a path is not legal, and this is the behaviour I see if I try to access `file://Users/robertbrignull` because it tries to load from the non-existent `Users` hostname. But both `file:///Users/robertbrignull` and `file:Users/robertbrignull` work as expected.